### PR TITLE
CSS manipulation + tests

### PR DESCRIPTION
**Pull request by bitpshr**
This PR includes `addClass`, `containsClass`, `removeClass`, `toggleClass`, and associated unit tests.

### CODE REVIEW

File: src/dom.ts [hunk 1]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+ * Removes all instances of one ore more CSS class names from an Element
+ *
+ * @param element The Element from which to remove CSS classes
+ * @param classes An array of string CSS classes to remove from the Element
+ *
+ * @example
+ * dom.removeClass(document.body, 'loading');
+ *
+ * @example
+ * dom.removeClas(document.body, 'loading', 'pending');
```
Comment (dylans) [inline src/dom.ts]: typo

File: src/dom.ts [hunk 2]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+	}
+}
+
+/**
+ * Toggles the presence of a CSS class name on an Element. An optional
+ * second parameter can be used to force class addition or removal.
+ *
+ * @param element The Element to add or remove classes to or from
+ * @param className The CSS class name add or remove
+ * @param force Forces either class addition if true or class removal if flase
```
Comment (dylans) [inline src/dom.ts]: typo

File: tests/unit/dom.ts [hunk 3]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+						dom.toggleClass(targetElement, 'test', false);
+						assert.notInclude(targetElement.className, 'test');
+					}
+				}
+			};
+		}
+
+		return {
+			HTMLElement: cssSuite(document.createElement('div')),
+			SVGElement: cssSuite(document.createElement('svg'))
```
Comment (dylans) [inline tests/unit/dom.ts]: Are we actually supporting SVG? I wouldn't think any of the methods we have currently would pass with SVG?

File: src/dom.ts [hunk 4]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+	if (/\s/.test(token)) {
+		throw new Error('String contains an invalid character');
+	}
+}
+
+/**
+ * Adds one or more CSS class names to an Element without duplication
+ *
+ * @param element The Element to which to add CSS classes
+ * @param classes An array of string CSS classes to add to the Element
```
Comment (kfranqueiro) [inline src/dom.ts]: I wouldn't say "an array" here since it's not actually an array, it's a rest argument.  I'd probably just say "one or more CSS class strings".

File: src/dom.ts [hunk 5]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+ *
+ * @example
+ * dom.addClass(document.body, 'loaded');
+ *
+ * @example
+ * dom.addClass(document.body, 'loaded', 'ready');
+ */
+export function addClass(element: Element, ...classes: string[]): void {
+	let targetElement = <any> element;
```
Comment (kfranqueiro) [inline src/dom.ts]: It might be worth noting https://github.com/Microsoft/TypeScript/issues/3220 in a comment above this to explain the `<any>` cast.

File: src/dom.ts [hunk 6]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+	if (!targetElement) {
+		return;
+	}
+	if (targetElement.classList) {
+		let classList: any = targetElement.classList;
+		classList.add.apply(classList, classes);
```
Comment (kfranqueiro) [inline src/dom.ts]: This is overly naive.  Not all browsers that support `classList` support multi-argument `add` and `remove` (this would fail in IE11, for example, only adding the first class).

We have a couple of options:

* I proposed has features for multi-argument add/remove and 2-argument toggle in the polyfill section of the dom proposal (obviously we're not doing polyfills right now so it would've been easy to overlook that).  It seems like it should be reasonable to add generic has tests for multi-argument support in these methods when they exist.  We would still need to test `targetElement.classList` directly on each call since IE doesn't support `classList` in SVG elements.
* The other option is to just go with the `else` branch below in all cases.

File: src/dom.ts [hunk 7]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+	}
+	else {
+		let newClasses: string[] = new Array();
```
Comment (kfranqueiro) [inline src/dom.ts]: Replace `new Array()` with `[]`

File: src/dom.ts [hunk 8]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+		for (let className in classes) {
```
Comment (kfranqueiro) [inline src/dom.ts]: I think you wanted `of` here, not `in`...

File: src/dom.ts [hunk 9]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+			validateClass(className);
+			if (!this.contains(targetElement, className)) {
```
Comment (kfranqueiro) [inline src/dom.ts]: What is `this`?  What is `contains`?  This should presumably just be `if (!containsClass(...))`.

Moreover, I'm wondering whether it'd be worth actually parsing the element's `className` into an object hash to make this more efficient for multiple-add.  Probably not initially, would have to benchmark it.

Please run the tests in IE9 to test this code path.  There are 24 test failures on this branch right now.

File: src/dom.ts [hunk 10]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+				newClasses.push(className);
+			}
+		}
+		if (newClasses.length > 0) {
+			targetElement.className += (' ' + newClasses.join(' '));
```
Comment (kfranqueiro) [inline src/dom.ts]: I'm nit-picking, but the initial space only needs to be added if `targetElement.className` is currently not empty.

File: src/dom.ts [hunk 11]
```diff
@@ -34,6 +78,30 @@ export function byId(id: string): HTMLElement {
 }
 
 /**
+ * Determines whether an ELement has a CSS class name
```
Comment (kfranqueiro) [inline src/dom.ts]: Typo (capital L in element)

File: src/dom.ts [hunk 12]
```diff
@@ -34,6 +78,30 @@ export function byId(id: string): HTMLElement {
...
+ *
+ * @param element The Element to which to check for a CSS class
```
Comment (kfranqueiro) [inline src/dom.ts]: Remove "to which"

File: src/dom.ts [hunk 13]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
...
+ * @example
+ * dom.addClass(document.body, 'loaded', 'ready');
+ */
+export function addClass(element: Element, ...classes: string[]): void {
+	let targetElement = <any> element;
+	if (!targetElement) {
+		return;
+	}
+	if (targetElement.classList) {
+		let classList: any = targetElement.classList;
```
Comment (kfranqueiro) [inline src/dom.ts]: This should be of type `DOMTokenList`

File: src/dom.ts [hunk 14]
```diff
@@ -34,6 +78,30 @@ export function byId(id: string): HTMLElement {
...
+ * @example
+ * var loaded = dom.containsClass(document.body, 'loaded');
+ */
+export function containsClass(element: Element, className: string): boolean {
+	let targetElement = <any> element;
+	if (!targetElement) {
+		return;
+	}
+	if (targetElement.classList) {
+		let classList: any = targetElement.classList;
```
Comment (kfranqueiro) [inline src/dom.ts]: This should be of type `DOMTokenList`

File: src/dom.ts [hunk 15]
```diff
@@ -34,6 +78,30 @@ export function byId(id: string): HTMLElement {
...
+		return classList.contains(className);
+	}
+	else {
+		validateClass(className);
+		return targetElement.className.indexOf(className) > -1;
```
Comment (kfranqueiro) [inline src/dom.ts]: This is overly naive.  For an element with a `className` of `foobar`, this would return true for `containsClass(element, 'foo')`.

I'd suggest adopting Dojo 1's implementation for this one.

File: src/dom.ts [hunk 16]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+ * dom.removeClas(document.body, 'loading', 'pending');
+ */
+export function removeClass(element: Element, ...classes: string[]): void {
+	let targetElement = <any> element;
+	if (!targetElement) {
+		return;
+	}
+	if (targetElement.classList) {
+		let classList: any = targetElement.classList;
+		classList.remove.apply(classList, classes);
```
Comment (kfranqueiro) [inline src/dom.ts]: Same remarks as `classList.add` apply here.

File: src/dom.ts [hunk 17]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+	}
+	else {
+		let oldClasses: string[] = targetElement.className.split(/\s+/);
+		let length = oldClasses.length;
+		for (let className in classes) {
```
Comment (kfranqueiro) [inline src/dom.ts]: `of`, not `in`

File: src/dom.ts [hunk 18]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+ * @param className The CSS class name add or remove
+ * @param force Forces either class addition if true or class removal if flase
+ *
+ * @example
+ * dom.toggleClass(button, 'active');
+ *
+ * @example
+ * dom.toggleClass(button, 'active', determineState());
+ */
+export function toggleClass(element: Element, className: string, force?: boolean): boolean {
```
Comment (kfranqueiro) [inline src/dom.ts]: It may actually be possible to get away with a default parameter here to simplify logic in the function.  e.g. `force: boolean = containsClass(element, className)`

File: src/dom.ts [hunk 19]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+	let targetElement = <any> element;
+	let forced = arguments.length > 2;
+	if (!targetElement) {
+		return;
+	}
+	if (targetElement.classList) {
+		let classList: any = targetElement.classList;
+		return classList.toggle.apply(classList, forced ? [className, force] : [className]);
```
Comment (kfranqueiro) [inline src/dom.ts]: Unfortunately, this is also overly naive.  Not all browsers support the `force` parameter (namely IE again).  See previous remarks RE `addClass`.

File: src/dom.ts [hunk 20]
```diff
@@ -140,3 +208,78 @@ export function remove(node: Node) {
...
+	}
+	else {
+		let contains = this.contains(element, className);
+		this[contains ? 'removeClass' : 'addClass'](element, className);
```
Comment (kfranqueiro) [inline src/dom.ts]: This doesn't seem to be accounting for `force` at all.  Assuming the idea above about a default parameter works, you should be able to just use `force` directly here (and for the return value).

File: tests/unit/dom.ts [hunk 21]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+		function cssSuite(element: Element): {} {
+			let targetElement = <any> element;
+			return {
+				addClass: {
+					afterEach() {
+						targetElement.className = '';
+					},
+
+					'add single class'() {
+						assert.notInclude(targetElement.className, 'test');
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: TIL `include` and `notInclude` works for strings.  Neat.

File: tests/unit/dom.ts [hunk 22]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+					'null class'() {
+						assert.doesNotThrow(function () {
+							dom.addClass(targetElement);
+						});
+					},
+
+					'existing class'() {
+						dom.addClass(targetElement, 'test');
+						dom.addClass(targetElement, 'test');
+						assert.lengthOf(targetElement.className.match(/test/), 1);
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: This will always be 1 because you left off the `g` flag.

Alternatively, you could just check `className` before and after the first `addClass` - it should be the same at both times.

File: tests/unit/dom.ts [hunk 23]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+						});
+					},
+
+					'invalid class'() {
+						assert.throws(function () {
+							dom.addClass(targetElement, 'te est');
+						});
+					},
+
+					'null class'() {
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: A better description for this would be "class not specified".

I was about to ask why passing a null class shouldn't throw, since the spec seems to say nothing about them, but does say that empty strings should throw (which I don't think we're doing at the moment?)

File: tests/unit/dom.ts [hunk 24]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+
+					'null node'() {
+						assert.doesNotThrow(function () {
+							dom.addClass(null, 'test');
+						});
+					},
+
+					'invalid class'() {
+						assert.throws(function () {
+							dom.addClass(targetElement, 'te est');
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: We should also probably be testing passing `null` or `''` here since those should throw too.

File: tests/unit/dom.ts [hunk 25]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+					},
+
+					'existing class'() {
+						dom.addClass(targetElement, 'test');
+						dom.addClass(targetElement, 'test');
+						assert.lengthOf(targetElement.className.match(/test/), 1);
+					}
+				},
+
+				containsClass: {
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: Add a test to this sub-suite for testing `contains` for a class that is a substring of a class that was actually added.

File: tests/unit/dom.ts [hunk 26]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+
+					'remove single class'() {
+						dom.addClass(targetElement, 'test');
+						assert.include(targetElement.className, 'test');
+						dom.removeClass(targetElement, 'test');
+						assert.notInclude(targetElement.className, 'test');
+					},
+
+					'remove multiple classes'() {
+						dom.addClass(targetElement, 'test1', 'test2');
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: Use `className` here to avoid "two wrongs make a right" false--positives in browsers that don't support multi-argument add/remove.

File: tests/unit/dom.ts [hunk 27]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+
+					'null node'() {
+						assert.doesNotThrow(function () {
+							dom.removeClass(null, 'test');
+						});
+					},
+
+					'invalid class'() {
+						assert.throws(function () {
+							dom.removeClass(targetElement, 'te est');
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: We should also probably be testing passing `null` or `''` here since those should throw too.

File: tests/unit/dom.ts [hunk 28]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+
+					'null node'() {
+						assert.doesNotThrow(function () {
+							dom.containsClass(null, 'test');
+						});
+					},
+
+					'invalid class'() {
+						assert.throws(function () {
+							dom.containsClass(targetElement, 'te est');
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: We should also probably be testing passing `null` or `''` here since those should throw too.

File: tests/unit/dom.ts [hunk 29]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+						});
+					},
+
+					'null class'() {
+						assert.doesNotThrow(function () {
+							dom.containsClass(targetElement, null);
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: This *should* throw, according to the spec...

I assume this test case isn't testing passing nothing because TS makes that impossible without suppressing a warning.  I'd vote to just remove this test case for this function then.

File: src/dom.ts [hunk 30]
```diff
@@ -20,6 +20,50 @@ for (const param in tagWrap) {
 	tw.post = '</' + tw.reverse().join('></') + '>';
 }
 
+function validateClass(token: string): void {
+	if (!token || token === '') {
```
Comment (kfranqueiro) [inline src/dom.ts]: This is repetitive.  `token === ''` will already satisfy `!token`.

Comment (kfranqueiro) [inline src/dom.ts]: Actually...  scratch that.  We should _only_ be checking `token === ''`.

The spec is ambiguous about what to do for null or undefined, but it appears that what actual implementations do is coerce to string first.  This causes `null` to result in `'null'` and `undefined` to result in `'undefined'`, etc.

As silly as that seems, if we want to be consistent with the behavior of native `classList`, we should consider coercing to string and returning the coerced value back to the caller.  (We should still keep the `string` type on the parameter too, but this will cover us for `null`).

File: tests/unit/dom.ts [hunk 31]
```diff
@@ -237,5 +237,169 @@ registerSuite({
...
+					'null class'() {
+						assert.doesNotThrow(function () {
+							dom.removeClass(targetElement);
+						});
+					},
+
+					'remove nonexistent class'() {
+						assert.doesNotThrow(function () {
+							dom.removeClass(targetElement, 'random');
+						});
```
Comment (kfranqueiro) [inline tests/unit/dom.ts]: It would probably also be a good idea to add some unrelated classes to the element and make sure they still exist after this (i.e. no adverse side effects).

I'd also probably just combine the multiple test below into this one (still in a separate `doesNotThrow`), and you might also add one that tries to remove one class that does exist and one class that doesn't at the same time.

File: tests/unit/dom.ts (hunk 3 above)
Comment (kfranqueiro) [inline tests/unit/dom.ts]: What methods are you thinking of?  I particularly poked around with SVG yesterday while talking to Paul about these APIs, and other than the fact that IE doesn't support native `classList` at all on them, they should work (they expose `className`).

Incidentally, if we didn't have to support IE9 *or* SVG, we could've saved some time here because we could always rely on native `classList`, we'd just have to call the native methods multiple times in browsers that don't support multi-argument add/remove, and would have to do some finagling for `toggle` in browsers that don't support the force argument.

Comment (dylans) [inline tests/unit/dom.ts]: I thought most of SVG required you to use things like createElementNS, and provide the namespace as a parameter... but maybe that's been fixed/simplified. That said, I was thinking this was for all DOM tests, not just the class manipulation subset. So I guess I'm more concerned about fromString and place being correct for an SVG environment, of the things we currently include. I also thought for getting and setting classes in SVG you had to use getAttribute/setAttribute rather than relying on element.className? Anyway, I guess my overall point is to make sure we are reviewing this with SVG support in mind.

Comment (bitpshr) [inline tests/unit/dom.ts]: This is a good point - should we circle back and make sure `fromString`, `place`, and the rest of dom also supports SVG?

Comment (kfranqueiro) [inline tests/unit/dom.ts]: Definitely good points, and I clearly underestimated the complexity here.  It appears that `className` is indeed different for SVG elements, so I suspect the SVG-related tests would still fail at least on IE which doesn't support `classList`...  and if we remove the `classList` code entirely to simplify, it may fail everywhere.

So I guess the first question indeed is, *do* we care about supporting SVG elements in these APIs or should we just KISS and declare that they only accept/return `HTMLElement`?  Maybe SVG concerns would be best deferred until we think about packages for gfx and charting?

File: tests/unit/dom.ts (hunk 23 above)
Comment (bitpshr) [inline tests/unit/dom.ts]: The `classList` API behaves different for an explicitly null class vs. no specified class. I updated the descriptions to be clearer, because your point about lack of clarity still stands.

File: tests/unit/dom.ts (hunk 29 above)
Comment (bitpshr) [inline tests/unit/dom.ts]: I updated all tests to be consistent with native implementations and to have clearer descriptions.

File: tests/unit/dom.ts (hunk 3 above)
Comment (bitpshr) [inline tests/unit/dom.ts]: For now, these tests are passing using a pure-`className` implementation. If we don't want to devote the time necessary to make _all_ `dom` methods work with SVG, I vote we punt it.

**Comment by bitpshr** [comment]
This should be ready for another review. Thanks!

File: tests/unit/dom.ts (hunk 3 above)
Comment (kfranqueiro) [inline tests/unit/dom.ts]: I think the reason these pass even for SVG is because the SVG element is created with `document.createElement`.  In this case, at least in Chrome where I've been poking around, `className` operates like it would for any DOM node.  However, SVG elements created with `document.createElementNS` - or authored directly within an HTML document - do not behave this way, and I imagine the tests would fail for that.

So yeah, I am thinking we should punt on this for now and create NS-aware equivalents when the time comes.

**Comment by kfranqueiro** [comment]
Merged via https://github.com/dojo/dom/commit/8d5def09070647b0719d8a9ad1c6ee8d822cebb2.  I removed SVG testing and made the functions expect HTMLElement, since this really doesn't support SVG elements (created via `createElementNS`).
